File: leetcode/1512.py
"""
Runtime: 36 ms, faster than 57.75% of Python3 online submissions for Number of Good Pairs.
Memory Usage: 14.2 MB, less than 44.43% of Python3 online submissions for Number of Good Pairs.
"""
from collections import Counter


class Solution:
    def numIdenticalPairs(self, nums: List[int]) -> int:
        c = Counter(nums)
        res = 0
        for v in c.values():
            res += (v - 1) * v // 2
        return sum(map(lambda x: (x - 1) * x // 2, [v for v in c.values()]))

File: leetcode/1512.cpp
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Number of Good Pairs.
// Memory Usage: 7.4 MB, less than 15.48% of C++ online submissions for Number of Good Pairs.
class Solution {
 public:
  int numIdenticalPairs(vector<int> &nums) {
    // measure the count
    map<int, int> cnt = {};
    for (auto n : nums) {
      cnt[n] = cnt.find(n) == cnt.end() ? 1 : cnt[n] + 1;
    }

    // for each n in nums; cnt[n] -= 1; res += cnt[n]
    int res = 0;
    for (auto n : nums) {
      res += (--cnt[n]);
    }

    return res;
  }
};
